File: bindings/ruby/tailwind.rb
require "fiddle"
require "fiddle/import"

module Tailwind
  extend Fiddle::Importer

  lib =
    if RUBY_PLATFORM =~ /mingw|mswin/
      "tailwind_styled_parser.dll"
    elsif RUBY_PLATFORM =~ /darwin/
      "libtailwind_styled_parser.dylib"
    else
      "libtailwind_styled_parser.so"
    end

  dlload File.expand_path("../../native/target/release/#{lib}", __dir__)

  extern "char* tailwind_compile(const char*)"
  extern "char* tailwind_compile_with_stats(const char*)"
  extern "void tailwind_free(char*)"
  extern "const char* tailwind_version()"
  extern "void tailwind_clear_cache()"

  class Compiler
    def compile(source)
      ptr = Tailwind.tailwind_compile(source)
      value = ptr.to_s
      Tailwind.tailwind_free(ptr)
      value
    end

    def compile_with_stats(source)
      ptr = Tailwind.tailwind_compile_with_stats(source)
      value = ptr.to_s
      Tailwind.tailwind_free(ptr)
      value
    end

    def version
      Tailwind.tailwind_version.to_s
    end

    def clear_cache
      Tailwind.tailwind_clear_cache
    end
  end
end

File: bindings/php/tailwind.php
<?php

final class TailwindCompiler {
    private FFI $ffi;

    public function __construct(?string $libraryPath = null) {
        if ($libraryPath === null) {
            $libraryPath = $this->defaultLibraryPath();
        }

        $this->ffi = FFI::cdef('
            char* tailwind_compile(const char* code);
            char* tailwind_compile_with_stats(const char* code);
            void tailwind_free(char* ptr);
            const char* tailwind_version(void);
            void tailwind_clear_cache(void);
        ', $libraryPath);
    }

    public function compile(string $source): string {
        $ptr = $this->ffi->tailwind_compile($source);
        try {
            return FFI::string($ptr);
        } finally {
            $this->ffi->tailwind_free($ptr);
        }
    }

    public function compileWithStats(string $source): string {
        $ptr = $this->ffi->tailwind_compile_with_stats($source);
        try {
            return FFI::string($ptr);
        } finally {
            $this->ffi->tailwind_free($ptr);
        }
    }

    public function version(): string {
        return FFI::string($this->ffi->tailwind_version());
    }

    public function clearCache(): void {
        $this->ffi->tailwind_clear_cache();
    }

    private function defaultLibraryPath(): string {
        $root = dirname(__DIR__, 2);
        if (PHP_OS_FAMILY === 'Windows') {
            return $root . '\\native\\target\\release\\tailwind_styled_parser.dll';
        }
        if (PHP_OS_FAMILY === 'Darwin') {
            return $root . '/native/target/release/libtailwind_styled_parser.dylib';
        }
        return $root . '/native/target/release/libtailwind_styled_parser.so';
    }
}
